### PR TITLE
Distinguish between local NVMEs and remote disks when disc controller type is NVMe

### PR DESCRIPTION
**Pull request by SRIKKANTH**

### CODE REVIEW

File: lisa/features/nvme.py [hunk 1]
```diff
@@ -73,6 +74,16 @@ def get_namespaces(self) -> List[str]:
...
         return self._node.tools[Nvmecli].get_namespaces()
 
+    def get_os_partition_namespace(self) -> str:
+        node_disk = self._node.features[Disk]
+        os_partition_namespace = ""
+        os_boot_partition = node_disk.get_os_boot_partition()
+        # Sample os_boot_partition when disc controller type is NVMe:
+        # name: /dev/nvme0n1p15, disk: nvme, mount_point: /boot/efi, type: vfat
+        if os_boot_partition:
+            os_partition_namespace = os_boot_partition.name.split("p")[0]
```
Comment (LiliDeng) [inline lisa/features/nvme.py]: use pattern to match

File: lisa/features/nvme.py [hunk 2]
```diff
@@ -73,6 +74,16 @@ def get_namespaces(self) -> List[str]:
     def get_namespaces_from_cli(self) -> List[str]:
         return self._node.tools[Nvmecli].get_namespaces()
 
+    def get_os_partition_namespace(self) -> str:
+        node_disk = self._node.features[Disk]
+        os_partition_namespace = ""
+        os_boot_partition = node_disk.get_os_boot_partition()
+        # Sample os_boot_partition when disc controller type is NVMe:
+        # name: /dev/nvme0n1p15, disk: nvme, mount_point: /boot/efi, type: vfat
```
Comment (squirrelsc) [inline lisa/features/nvme.py:85]: The comments haven't specified what's the "namespace". Is it a general concept of disk controller?

File: lisa/features/nvme.py [hunk 3]
```diff
@@ -85,6 +96,18 @@ def get_devices_from_lspci(self) -> List[PciDevice]:
...
+    def get_raw_nvme_disks(self) -> List[str]:
+        nvme_namespaces = self.get_namespaces()
+        # With disk controller type NVMe, OS disk and all remote disks appear as NVMes.
+        # They should be removed from the list of namespaces for NVMe tests as they are
+        # not actual NVMe devices.
+        # OS disk and all remote disks are connected to the same NVMe controller.
+        # Excluding the OS disk's namespace from the list of namespaces will exclude
+        # all remote disks.
+        get_os_partition_namespace = self.get_os_partition_namespace()
+        nvme_namespaces.remove(get_os_partition_namespace)
```
Comment (squirrelsc) [inline lisa/features/nvme.py]: After removed, is the os disk included yet? I cannot get the information from above comments.

File: lisa/features/nvme.py [hunk 4]
```diff
@@ -85,6 +96,18 @@ def get_devices_from_lspci(self) -> List[PciDevice]:
...
 
+    def get_raw_nvme_disks(self) -> List[str]:
+        nvme_namespaces = self.get_namespaces()
+        # With disk controller type NVMe, OS disk and all remote disks appear as NVMes.
+        # They should be removed from the list of namespaces for NVMe tests as they are
+        # not actual NVMe devices.
+        # OS disk and all remote disks are connected to the same NVMe controller.
+        # Excluding the OS disk's namespace from the list of namespaces will exclude
+        # all remote disks.
+        get_os_partition_namespace = self.get_os_partition_namespace()
```
Comment (squirrelsc) [inline lisa/features/nvme.py]: The variable name should be like `os_partition_namespace` without get.

File: lisa/sut_orchestrator/azure/features.py [hunk 5]
```diff
@@ -1545,15 +1545,31 @@ def get_raw_data_disks(self) -> List[str]:
...
-        ):
+        # disk_controller_type == NVME
+        node_disk = self._node.features[Disk]
+        if node_disk.get_os_disk_controller_type() == schema.DiskControllerType.NVME:
+            # Getting OS disk nvme namespace and disk controller used by OS disk.
+            # Sample os_boot_partition:
+            # name: /dev/nvme0n1p15, disk: nvme, mount_point: /boot/efi, type: vfat
+            os_boot_partition = node_disk.get_os_boot_partition()
+            if os_boot_partition:
+                os_disk_namespace = os_boot_partition.name.split("p")[0]
```
Comment (squirrelsc) [inline lisa/sut_orchestrator/azure/features.py]: use regexp

File: lisa/features/nvme.py (hunk 2 above)
Comment (SRIKKANTH) [reply lisa/features/nvme.py:85]: namespace is a concept of NVME. 
"In NVMe® technology, a namespace is a collection of logical block addresses (LBA) accessible to host software. A namespace ID (NSID) is an identifier used by a controller to provide access to a namespace. A namespace is not the physical isolation of blocks, rather the isolation of logical blocks addressable by the host software.

In Linux, namespaces appear each with a unique identifier in devices; /dev/nvme0n1 is looking at controller 0 and namespace 1)."

Please see: https://nvmexpress.org/resource/nvme-namespaces/

Comment (squirrelsc) [reply lisa/features/nvme.py:85]: The method name should be `get_os_nvme_namespace`.

File: lisa/features/nvme.py (hunk 3 above)
Comment (SRIKKANTH) [reply lisa/features/nvme.py]: Updated comments please check once again.

File: lisa/features/nvme.py (hunk 2 above)
Comment (SRIKKANTH) [reply lisa/features/nvme.py:85]: Changed accordingly.

File: lisa/features/nvme.py (hunk 4 above)
Comment (SRIKKANTH) [reply lisa/features/nvme.py]: Fixed

File: lisa/features/nvme.py [hunk 6]
```diff
@@ -35,6 +36,8 @@ class Nvme(Feature):
         r".*(?P<namespace>/dev/nvme[0-9]ns[0-9]+$)", re.MULTILINE
     )
 
+    NVME_NAMESPACE_PATTERN = re.compile(r"/dev/nvme[0-9]+n[0-9]+", re.M)
```
Comment (squirrelsc) [inline lisa/features/nvme.py:40]: Add examples above the regexp.

Comment (SRIKKANTH) [reply lisa/features/nvme.py:40]: Added example.

File: lisa/features/nvme.py (hunk 1 above)
Comment (SRIKKANTH) [reply lisa/features/nvme.py]: Fixed

**Comment by squirrelsc** [comment]
@LiliDeng LGTM

**Comment by LiliDeng** [comment]
@SRIKKANTH please run basic functional and regression testing, I hit below error when run nvme functional tests.
```Nvme.verify_nvme_basic: FAILED   failed. ValueError: list.remove(x): x not in list```

**Review by LiliDeng — APPROVED** [review]

**Pull request merged**